File: day:40/twoSumDesign.js
/**
 * Initialize your data structure here.
 */
var TwoSum = function() {
    this.hash = {}
};

/**
 * Add the number to an internal data structure.. 
 * @param {number} number
 * @return {void}
 */
TwoSum.prototype.add = function(number) {
    if(this.hash[number]) {
        this.hash[number] = this.hash[number] + 1;
    }
    else {
        this.hash[number] = 1;
    }
};

/**
 * Find if there exists any pair of numbers which sum is equal to the value. 
 * @param {number} value
 * @return {boolean}
 */
TwoSum.prototype.find = function(value) {
    for(const key of Object.keys(this.hash)) {
        const intKey = parseInt(key);
        const diff = value - intKey;
        if(intKey === diff) {
            if(this.hash[key] > 1) {
                return true;
            }
        }
        else {
            if(this.hash[(diff).toString()]) {
                return true;
            }
        }
    }
    return false;
};

/** 
 * Your TwoSum object will be instantiated and called as such:
 * var obj = new TwoSum()
 * obj.add(number)
 * var param_2 = obj.find(value)
 */

File: day:40/twosumdesign.py
class TwoSum:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.hash = {}

    def add(self, number: int) -> None:
        """
        Add the number to an internal data structure..
        """
        if(number in self.hash):
            self.hash[number] += 1;
        else:
            self.hash[number] = 1;

    def find(self, value: int) -> bool:
        """
        Find if there exists any pair of numbers which sum is equal to the value.
        """
        for key in self.hash.keys():
            diff = value - key;
            if(key == diff):
                if(self.hash[key] > 1):
                    return True;
            else:
                if( diff in self.hash):
                    return True;
        return False;
        


# Your TwoSum object will be instantiated and called as such:
# obj = TwoSum()
# obj.add(number)
# param_2 = obj.find(value)
